File: geekbang/algorithm-models/1.1-单调栈模板/largestRectangleArea.go
package main

// Go
// LeetCode 84 柱状图中最大的矩形

type Rect struct {
	Height int
	Width  int
}

func largestRectangleArea(heights []int) int {
	heights = append(heights, 0) // 帮助我们在最后把栈清空
	var s []Rect
	ans := 0
	// 每个柱子入栈、出栈各一次，2n=O(n)
	// 第一步：for 每个元素
	for _, h := range heights {
		accumulatedWidth := 0
		// 第二步：while (栈顶不满足高度单调性) 累加宽度，出栈
		for len(s) > 0 && s[len(s)-1].Height >= h {
			accumulatedWidth += s[len(s)-1].Width
			if ans < accumulatedWidth*s[len(s)-1].Height {
				ans = accumulatedWidth * s[len(s)-1].Height
			}
			s = s[:len(s)-1]
		}
		// 第三步：新元素入栈
		s = append(s, Rect{h, accumulatedWidth + 1})
	}
	return ans
}

File: geekbang/algorithm-models/1.1-单调栈模板/largestRectangleArea.cpp
// C/C++
// LeetCode 84 柱状图中最大的矩形
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        heights.push_back(0); // 帮助我们在最后把栈清空
        stack<Rect> s;
        int ans = 0;
        // 每个柱子入栈、出栈各一次，2n=O(n)
        // 第一步：for 每个元素
        for (int h : heights) {
            int accumulated_width = 0;
            // 第二步：while (栈顶不满足高度单调性) 累加宽度，出栈
            while (!s.empty() && s.top().height >= h) {
                accumulated_width += s.top().width;
                ans = max(ans, accumulated_width * s.top().height);
                s.pop();
            }
            // 第三步：新元素入栈
            s.push({h, accumulated_width + 1});
        }
        return ans;
    }


private:
    struct Rect {
        int height;
        int width;
    };
};
